File: 백준/Silver/1213. 팰린드롬 만들기/팰린드롬 만들기.cc
#include <bits/stdc++.h>
using namespace std;

string str;
int arr[200], flag;
string answer;
char mid;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);
	
	cin >> str;

	for (char c : str) arr[c]++;

	for (int i = 'Z'; i >= 'A'; --i) {
		// 개수가 홀수인 경우
		if (arr[i] & 1 == 1) {
			mid = char(i);
			flag++;
			arr[i]--;
		}

		// 홀수가 2개 이상 나오면 팰린드롬 만들 수 없음
		if (flag == 2)
			break;

		for (int j = 0; j < arr[i]; j += 2) {
			answer = char(i) + answer + char(i);
		}
	}

	if (mid)
		answer.insert(answer.begin() + answer.size() / 2, mid);

	if (flag == 2) {
		cout << "I'm Sorry Hansoo";
	}
	else {
		cout << answer;
	}

	return 0;
}

File: 백준/Silver/1213. 팰린드롬 만들기/팰린드롬 만들기.py
import sys
input = sys.stdin.readline

text = input().strip()

arr = [0] * 100

A = ord('A')

for c in text:
    arr[ord(c) - A] += 1

flag = 0
mid = ""
ans = ""

for i in range(ord('Z') - ord('A'), -1, -1):
    if arr[i] % 2 == 1:
        flag += 1
        mid = chr(i + ord('A'))
        arr[i] -= 1

    if flag >= 2:
        break

    for _ in range(arr[i] // 2):
        ans = chr(i + ord('A')) + ans + chr(i + ord('A'))

if len(mid) > 0:
    insert_idx = len(ans) // 2
    ans = ans[:insert_idx] + mid + ans[insert_idx:]

if flag >= 2:
    print("I\'m Sorry Hansoo")
else:
    print(ans)